### ISSUE TITLE
Both "High" and "Low" sound options quality is poor.

### ISSUE THREAD

**Issue by JokerOmega1** [JACoders/OpenJK#798]
Quality on both High and Low sound options is poor, the scratchiness literally hurts the ears.
EAX enhanced can help as an alternative but I'd really like to see the default direct sound feature be placed back in over the one currently in place.

Even the hidden very high setting at 44 is just as scratchy and poor!

**Comment by Razish** [comment]
Try adjusting `s_sdlSpeed` (44100?) - I expect the menus are adjusting the old `s_khz` cvar

**Comment by ensiform** [comment]
Razor is correct.

**Comment by JokerOmega1** [comment]
Yeah no matter whether I change from high to low it stays at 44100 (according to the console that is), and the result is scratchy audio, even when I use the console command s_khz cvar 44.

I can try the s_sdlSpeed command, but I am unsure as to what is the best value to set it too, I mean it would be useless setting to 44100 if that is the default anyway which is causing the scratchy sound issue.

**Comment by ensiform** [comment]
What exactly sounds bad at 44?  Everything?

**Comment by JokerOmega1** [comment]
Basically all the sound effects, just the overall sound in general sounds scratchy, and not to exaggerate here but listening to it for just for a little while hurts my ears. I never had this problem with the vanilla game's audio, in fact it sounded just fine on the high setting.

This is what is stated on my OpenJK SP console on startup at High setting:
SDL Init (SLD_INIT_AUDIO) ... OK
SDL audio driver is "xaudio2"
SLD AudioSpec:
Format: Audio S16LSB
Freq: 44100
Sample: 1024
Channels: 2
Starting SDL audio callback.
SDL audio Initialized.

**Comment by ensiform** [comment]
On mine it says directsound.

**Comment by JokerOmega1** [comment]
Windows 7 x64

**Comment by ensiform** [comment]
Windows 8 and 10 both have had this for me:

```
------- sound initialization -------
SDL_Init( SDL_INIT_AUDIO )... OK
SDL audio driver is "directsound".
SDL_AudioSpec:
  Format:   AUDIO_S16LSB
  Freq:     44100
  Samples:  1024
  Channels: 2
Starting SDL audio callback...
SDL audio initialized.
```

**Comment by JokerOmega1** [comment]
Well geee that really is weird, I sure wish I could get OpenJK to do that too har har.

Really though, I'd like nothing more than to recommend OpenJK (in fact I already have under a mod that I have released that requires OpenJK to bypass the transform space for goul2), but this is just off putting. Don't get me wrong, I have nothing but appreciation for what you guys are doing... But there isn't really much I can do on my end to fix this, as far as I am aware of.

**Comment by Razish** [comment]
Could you upload a recording so we know what you're hearing?

**Comment by JokerOmega1** [comment]
I really don't think that is going to make much difference, given that I'd have to take the time to do a comparison with OpenJK and Vanilla JKA, plus difference of the sound transfer to the video as compared to the game itself.

It is basically like listening to a 64 kbps MP3.

If I could just get the directsound to activate as it should then I really don't think there would be a problem.

**Comment by ensiform** [comment]
XAudio2 support was disabled for other reasons in 2.0.4, so you'll have to wait until its switched.  Typically it would be possible to use the provided binary from the SDL site but with 2015 VS that doesn't work with older CRT versions correctly, and some of the SDL ABI changed in 2.0.4 as well I think.

**Comment by JokerOmega1** [comment]
Hmmm I am not quite sure how to do that, otherwise if need be I can wait for the next OpenJK release, if that has the SDL upgrade of course.

You've been very helpful on this issue, so thank you.

**Comment by ensiform** [comment]
Not sure when the upgrade will happen yet.

**Comment by JokerOmega1** [comment]
How complicated would you say it is setting the SDL_AUDIODRIVER in the enviroment variables, is this something done in the game console or through the PC settings?

**Comment by ensiform** [comment]
Through the system settings by going to the this pc's settings or w/e its called on 7 still. And under advanced system settings navigation pane and then environment variables.

You'd add a new variable under the user variables for <your username> called SDL_AUDIODRIVER with the value of directsound with no spaces or quotes

**Comment by Razish** [comment]
Alternatively, just from the game from `cmd.exe` with `SDL_AUDIODRIVER=directsound ./openjk.x86.exe +set fs_game "OpenJK"`

**Comment by JokerOmega1** [comment]
The environment variables method worked in the sense that the game started directsound instead of xaudio2, which was nice, only it made no difference when it came to the sound quality.

It just must be the way it has been coded into OpenJK, I mean in the vanilla game you can clearly tell a difference between low and high sound quality, with high sounding just fine, yet with OpenJK low and high options sound just the same in quality.

I think the main difference is the EAX sound option, in the vanilla game the EAX option sounded alright but was overall buggy with initiating sound effects and such. With OpenJK the EAX option seems to be fixed in the sense that sound effects now initiate as they should and the sound is great quality, except with this version there seems to be issues with imbalance in sound effects volume, for example on Korriban the background torches drown out every other sound, making even voices difficult to hear (and that is with sound effects bellow half and voices on full).

Here is my summation:

Vanilla JKA:
- Low sound option is how it sounds, low sound quality.
- High sound option is as it sounds, high sound quality and works just fine.
- EAX sound option is buggy, with sound effects failing to play at certain times.

OpenJK:
- Low sound option sounds scratchy.
- High sound option sounds exactly like low sound option.
- EAX sound option has great audio quality but terrible imbalance in audio volume.

Eh I can't help but feel disappointed, I was happy with vanilla JKA's high sound option, but I had issues with the game's ghoul2 transform limitation (crucial for single player model mods). OpenJK fixed the ghoul2 transform limitation, but now I have issues with sound.

**Comment by ensiform** [comment]
We've already told you the sound options in OpenJK do nothing in the menu.  They point to a dead cvar, unused cvar.

**Comment by JokerOmega1** [comment]
You've already told me?! Not that I can see, and if you have then you haven't made it at all clear.

All I can tell you is that the sound quality is nowhere near as good as vanilla JKA on my end. Hell I am not even sure how I am supposed to change the sound options in OpenJK if it isn't done through the menu, perhaps there's a console command that resets the OpenJK sound option, perhaps you could share that, or do I have to delete the OpenJK config and create a new one?

**Comment by ensiform** [comment]
Wasn't it mentioned earlier that s_khz is the old cvar?  It is unused therefore the option for sound quality does nothing which is fine so long as the other methods change quality.  And I'm not convinced its the sound quality causing issue.

**Comment by Razish** [comment]
Indeed, `s_khz` was the cvar used by basejka (and consequently the menus) measured in kHz
We now use `s_sdlSpeed` measured in hz, but the menus have not been updated because we legally can't ship OpenJK with modified menus, nor are we at an official release stage/1.0

**Comment by xycaleth** [comment]
I agree with @JokerOmega1 that with 44khz, the audio does sound more scratchy. It was noticeable for me in vanilla JKA as well, when I set s_khz from 22 (high quality) to 44 (not accessible except via the console).

IIRC, I had an onboard audio chip.

**Comment by ensiform** [comment]
I have an on board audio chip as well and don't really notice an issue per-se.

**Comment by xycaleth** [comment]
@JokerOmega1 Does OpenJK sound better if you set `s_sdlSpeed` to `22050`? This should be equivalent to the 'high quality' sound setting.

**Comment by ensiform** [comment]
22 would be lower than the highest setting in JKA ( the original default value )

**Comment by xycaleth** [comment]
Setting the sound quality to 'high' through the menu sets `s_khz` to 22. Low quality is 11. There was never an option to set it to 44 through the menus.

**Comment by ensiform** [comment]
I'm running around on a map right now with directsound, and the sdl equiv of 44 and it sounds great, including high rof weapons like repeater.

**Comment by JokerOmega1** [comment]
Oh wow, thanks xycaleth that did the trick, setting s_sdlSpeed to 22050 and then hitting /snd_restart fixed the sound, now it sounds just like it does in Vanilla's high setting.

**Comment by Dragonji** [comment]
BTW, don't you think it'd be better if s_sdl* cvars were renamed back to match vanillas so there's compatibility with original menus in this regard?

For example, they [did this in ET Legacy](https://github.com/etlegacy/etlegacy/commit/a54d2d7fa443fc7be041858d82ad017083648997) and I think it was a good change.

**Comment by Razish** [comment]
I'm not sure, I dislike the hackiness of correcting those values

**Comment by Dragonji** [comment]
Nobody said it must be done the same way, I just gave an example.

**Comment by ensiform** [comment]
I actually don't think it is.  I think it would be better if s_khz just set the sdl value if needed.

**Comment by Razish** [comment]
But only if modified before each sound restart, so the menu value matches it but we can still tweak `s_sdlSpeed`. This requires `s_kHz` to still technically exist

**Comment by Jon-guy30** [comment]
What's up with the sound on Windows 11? I downloaded openJK from here https://github.com/JACoders/OpenJK/releases/tag/latest whereas before I used to download from here: https://builds.openjk.org/

The audio lags very badly in the menus, when I played last time on Windows 10 everything was flawless. I have audio set to seta s_kHz "44".  Can one fix the audio in the openJK? Previously I only had this issue https://github.com/JACoders/OpenJK/issues/897 which still isn't fixed. Now that laggy sound is EVERYWHERE! I didn't bother to check the game proper, if everything else is OK. First impressions and all...

Sigh.

**Comment by taysta** [comment]
> What's up with the sound on Windows 11? I downloaded openJK from here https://github.com/JACoders/OpenJK/releases/tag/latest whereas before I used to download from here: https://builds.openjk.org/
> 
> The audio lags very badly in the menus, when I played last time on Windows 10 everything was flawless. I have audio set to seta s_kHz "44". Can one fix the audio in the openJK? Previously I only had this issue #897 which still isn't fixed. Now that laggy sound is EVERYWHERE! I didn't bother to check the game proper, if everything else is OK. First impressions and all...
> 
> Sigh.

Were you able to try the fix suggested here? https://github.com/JACoders/OpenJK/issues/798#issuecomment-175388305

**Comment by Jon-guy30** [comment]
I haven't yet tried that, I'll do it once I get the chance.

**Comment by Jon-guy30** [comment]
For a brief moment, I started a new game, my assumption was actually incorrect. it's only the menus where the sound is broken, the in-game and cutscene including the opening crawl didn't have any problems.

I also have the latest Realtek audio driver for my computer. What gives?

**Comment by ensiform** [comment]
OpenJK hasn't used s_sdlSpeed for ~6 years according to the commit history.

**Comment by taysta** [comment]
> For a brief moment, I started a new game, my assumption was actually incorrect. it's only the menus where the sound is broken, the in-game sound and cutscene audio including the opening crawl didn't have any problems.
> 
> I also have the latest Realtek audio driver for my computer. What gives? Should I downgrade openJK?
> 
> I also have an error, s_sdlSpeed command not found...
> 
> Screenshot:
> 
> ![Jedi Academy](https://private-user-images.githubusercontent.com/10241225/305989139-6dd15fc8-c1ec-49d9-8b9f-cba822d5002c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0MDE0NzgsIm5iZiI6MTcwODQwMTE3OCwicGF0aCI6Ii8xMDI0MTIyNS8zMDU5ODkxMzktNmRkMTVmYzgtYzFlYy00OWQ5LThiOWYtY2JhODIyZDUwMDJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIwVDAzNTI1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg4ZDlkYzFmMmZiOWZmMzQyNzUxZjliZWVjNTBlODliZjdhMDZiODI5MTExMmE4NWQ5MzNmZmVkOWI2Yzk4MmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.TZ6nBYS94Py-jqKTdQ8d59t7i44Lgs01sHHqHuAofcE)
> 
> Did further testing. I downloaded the latest build from https://builds.openjk.org/ which appears to be from fall 2023, and that one has correct audio in menus. The build from https://github.com/JACoders/OpenJK/releases/tag/latest which I downloaded yesterday is broken.

Is this a 64-bit build? The old builds site only provides 32-bit I believe, and EAX only works on 32-bit. Can I just confirm this isn't the case here?

**Comment by Jon-guy30** [comment]
Yeah, it's the x64 build.

I seem to recall EAX was turned off. I'll have to check that today.

Isn't EAX supposed to be disabled when using the "Very high" 44kHz audio setting?

**Comment by ensiform** [comment]
Eax doesn't function in x64 builds currently no matter what you see in the menu.

**Comment by Jon-guy30** [comment]
I don't think this is related to EAX.

Screenshots

x64: 
![JKA x64](https://github.com/JACoders/OpenJK/assets/10241225/66aa3e6a-4089-4ec7-8b8e-fa45ed07dd34)

x86: 
![JKA x86](https://github.com/JACoders/OpenJK/assets/10241225/6a114227-0e09-41fe-a612-b5ddd9f6ba81)

As you see, in both of them EAX is turned off by default (as it is also in the retail non-openJK JKA). The x86 build one can turn on EAX, in the x64 build it's not even an option...

Besides, using EAX disables 44Khz, I don't want that.

PCgamingwiki: "Using OpenAL may limit the music sampling rate to 22 kHz, regardless of snd_kHz setting" Link: https://www.pcgamingwiki.com/wiki/Star_Wars:_Jedi_Knight_-_Jedi_Academy#Restore_surround_sound_and_EAX

You see, I want 44Khz audio, not EAX. 

Off topic: Can I switch between openJK versions w/o uninstalling the game all the time? Thanks.

**Comment by ensiform** [comment]
Why would you need to uninstall the game all the time?

**Comment by Jon-guy30** [comment]
Because I don't know how to add different versions of openjk to the game, I tried and it gave me some kind of unhandled exception error when I tried launching the executable. 

Also, what OS are you yourself using if you don't have broken SFX? Windows 10?

**Comment by ensiform** [comment]
s_khz is the command that is used, I use Linux. There is no EAX on Linux and there is only the SDL/dma backend.

Sounds like you are probably running into issues with not keeping separate SDL dlls for 32-bit and 64-bit, which can simply be swapped out easier than a full reinstall of the game.

**Comment by Jon-guy30** [comment]
I'll try s_kHz tomorrow. I'll edit this post if it works. It's pretty late now.

Stay tuned.

**Comment by ensiform** [comment]
It's what the menu setting is using.

**Comment by Jon-guy30** [comment]
I now know what the problem is. No, I don't have a solution. It's in the menus, the sound is OK, but as soon as I hit ANY of the menu buttons the sound cuts off, i.e. pressing tab to look at mission objectives, audio cuts, to save the game, audio cuts, mission selection screen the ambient track cuts off in single player, Funniest thing I found was hitting "next" while Kyle or Luke gives you the mission briefing hovering over the "Next" button to go to the weapon selection screen the aduio cuts off and returns. "Th... is... shoul.........d...... be a total blu........e milk run kid, it..... 'll...... give you a chanc.....e...." Kyle's dialogue in the droid rescue mission. XD
Of course in the weapon selection screen the audio cuts off choosing weapons and force upgrades!!!

As long as you DON'T interact with the menus everything's fine and dandy, but as soon as you open the menu or interact with the buttons, the audio is turned off. It's very jarring and immersion breaking. It's not an audio quality problem, it's... something else.

Like I said, OLDER openjk revisions like the latest one from the builds site i A-OK!

Could you perhaps test this if it happens to you? Open singleplayer, easiest way to reproduce this if my bug is legit is to hit tab to look a mission objective, or just interact with the menus in a meaningful manner like, I don't know, saving the game.

Should I open a new thread, or what should I do?
